File: core/src/main/java/com/emarsys/core/request/RequestTask.java
package com.emarsys.core.request;

import android.os.AsyncTask;

import com.emarsys.core.CoreCompletionHandler;
import com.emarsys.core.Mapper;
import com.emarsys.core.connection.ConnectionProvider;
import com.emarsys.core.provider.timestamp.TimestampProvider;
import com.emarsys.core.request.model.RequestMethod;
import com.emarsys.core.request.model.RequestModel;
import com.emarsys.core.response.ResponseHandlersProcessor;
import com.emarsys.core.response.ResponseModel;
import com.emarsys.core.util.Assert;
import com.emarsys.core.util.JsonUtils;
import com.emarsys.core.util.log.Logger;
import com.emarsys.core.util.log.entry.RequestLog;

import java.io.BufferedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;

import javax.net.ssl.HttpsURLConnection;

import kotlin.text.Charsets;


public class RequestTask extends AsyncTask<Void, Long, Void> {
    private static final int TIMEOUT = 30_000;

    private final RequestModel requestModel;
    private final CoreCompletionHandler handler;
    private final ConnectionProvider connectionProvider;
    private final ResponseHandlersProcessor responseHandlersProcessor;
    private final List<Mapper<RequestModel, RequestModel>> requestModelMappers;
    private TimestampProvider timestampProvider;

    private ResponseModel responseModel;
    private Exception exception;

    public RequestTask(
            RequestModel requestModel,
            CoreCompletionHandler handler,
            ConnectionProvider connectionProvider,
            TimestampProvider timestampProvider,
            ResponseHandlersProcessor responseHandlersProcessor,
            List<Mapper<RequestModel, RequestModel>> requestModelMappers) {
        Assert.notNull(requestModel, "RequestModel must not be null!");
        Assert.notNull(handler, "CoreCompletionHandler must not be null!");
        Assert.notNull(connectionProvider, "ConnectionProvider must not be null!");
        Assert.notNull(timestampProvider, "TimestampProvider must not be null!");
        Assert.notNull(responseHandlersProcessor, "ResponseHandlersProcessor must not be null!");
        Assert.notNull(requestModelMappers, "RequestModelMappers must not be null!");

        this.requestModel = requestModel;
        this.handler = handler;
        this.connectionProvider = connectionProvider;
        this.timestampProvider = timestampProvider;
        this.responseHandlersProcessor = responseHandlersProcessor;
        this.requestModelMappers = requestModelMappers;
    }

    @Override
    protected Void doInBackground(Void... params) {
        long dbEnd = timestampProvider.provideTimestamp();

        HttpsURLConnection connection = null;
        try {
            RequestModel updatedRequestModel = mapRequestModel(requestModel);

            connection = connectionProvider.provideConnection(updatedRequestModel);

            initializeConnection(connection, updatedRequestModel);
            connection.setConnectTimeout(20_000);
            connection.connect();
            sendBody(connection, updatedRequestModel);
            responseModel = readResponse(connection);
            Logger.info(new RequestLog(responseModel, dbEnd));
        } catch (Exception e) {
            exception = e;
        } finally {
            if (connection != null) {
                connection.disconnect();
            }
        }
        return null;
    }

    @Override
    protected void onPostExecute(Void result) {
        if (exception != null) {
            handler.onError(requestModel.getId(), exception);
        } else if (responseModel != null) {
            responseHandlersProcessor.process(responseModel);
            if (isStatusCodeOK(responseModel.getStatusCode())) {
                handler.onSuccess(requestModel.getId(), responseModel);
            } else {
                handler.onError(requestModel.getId(), responseModel);
            }
        }
    }

    public RequestModel getRequestModel() {
        return requestModel;
    }

    public CoreCompletionHandler getHandler() {
        return handler;
    }

    private void initializeConnection(HttpsURLConnection connection, RequestModel model) throws IOException {
        connection.setRequestMethod(model.getMethod().name());
        setHeaders(connection, model.getHeaders());
        connection.setConnectTimeout(TIMEOUT);
        if (model.getMethod() != RequestMethod.GET && model.getPayload() != null) {
            connection.setDoOutput(true);
        }
    }

    private void setHeaders(HttpsURLConnection connection, Map<String, String> headers) {
        for (Map.Entry<String, String> entry : headers.entrySet()) {
            String key = entry.getKey();
            String value = entry.getValue();
            connection.setRequestProperty(key, value);
        }
    }

    private void sendBody(HttpsURLConnection connection, RequestModel model) throws IOException {
        if (model.getPayload() != null) {
            byte[] payload = JsonUtils.fromMap(model.getPayload()).toString().getBytes(StandardCharsets.UTF_8);
            BufferedOutputStream writer = new BufferedOutputStream(connection.getOutputStream());
            writer.write(payload);
            writer.close();
        }
    }

    private ResponseModel readResponse(HttpsURLConnection connection) throws IOException {
        int statusCode = connection.getResponseCode();
        String message = connection.getResponseMessage();
        Map<String, List<String>> headers = connection.getHeaderFields();
        byte[] bodyBytes = readByteArray(connection);
        return new ResponseModel.Builder(timestampProvider)
                .statusCode(statusCode)
                .message(message)
                .headers(headers)
                .body(new String(bodyBytes, Charsets.UTF_8))
                .bytes(bodyBytes)
                .requestModel(requestModel)
                .build();
    }

    private byte[] readByteArray(HttpsURLConnection connection) throws IOException {
        int responseCode = connection.getResponseCode();
        InputStream inputStream;
        if (isStatusCodeOK(responseCode)) {
            inputStream = connection.getInputStream();
        } else {
            inputStream = connection.getErrorStream();
        }
        ByteArrayOutputStream buffer = new ByteArrayOutputStream();

        int nRead;
        byte[] data = new byte[1024];
        while ((nRead = inputStream.read(data, 0, data.length)) != -1) {
            buffer.write(data, 0, nRead);
        }

        buffer.flush();
        return buffer.toByteArray();

    }

    private boolean isStatusCodeOK(int responseCode) {
        return 200 <= responseCode && responseCode < 300;
    }

    private RequestModel mapRequestModel(RequestModel requestModel) {
        RequestModel updatedRequestModel = requestModel;

        for (Mapper<RequestModel, RequestModel> mapper : requestModelMappers) {
            updatedRequestModel = mapper.map(updatedRequestModel);
        }

        return updatedRequestModel;
    }
}

File: core/src/main/java/com/emarsys/core/request/RequestTask.kt
package com.emarsys.core.request

import com.emarsys.core.Mockable
import com.emarsys.core.api.result.Try
import com.emarsys.core.connection.ConnectionProvider
import com.emarsys.core.provider.timestamp.TimestampProvider
import com.emarsys.core.request.model.RequestMethod
import com.emarsys.core.request.model.RequestModel
import com.emarsys.core.response.ResponseModel
import com.emarsys.core.util.JsonUtils.fromMap
import com.emarsys.core.util.filterNotNull
import com.emarsys.core.util.log.Logger.Companion.debug
import com.emarsys.core.util.log.Logger.Companion.info
import com.emarsys.core.util.log.entry.RequestLog
import java.io.BufferedOutputStream
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import javax.net.ssl.HttpsURLConnection

@Mockable
open class RequestTask(
    private val requestModel: RequestModel,
    private val connectionProvider: ConnectionProvider,
    private val timestampProvider: TimestampProvider
) {

    companion object {
        private const val TIMEOUT = 30000
    }

    open fun execute(): Try<ResponseModel> {
        var responseModel: ResponseModel? = null
        var exception: Exception? = null
        val dbEnd = timestampProvider.provideTimestamp()
        var connection: HttpsURLConnection? = null
        try {
            connection = connectionProvider.provideConnection(requestModel)
            initializeConnection(connection, requestModel)
            connection.connectTimeout = 20000
            connection.connect()

            sendBody(connection, requestModel)
            responseModel = readResponse(connection)

            debug(RequestLog(responseModel, dbEnd, requestModel))
            info(RequestLog(responseModel, dbEnd))

        } catch (e: Exception) {
            exception = e
        } finally {
            connection?.disconnect()
        }
        return Try(responseModel, exception)
    }

    private fun initializeConnection(connection: HttpsURLConnection, model: RequestModel) {
        connection.requestMethod = model.method.name
        setHeaders(connection, model.headers)
        connection.connectTimeout = TIMEOUT
        if (model.method != RequestMethod.GET && model.payload != null) {
            connection.doOutput = true
        }
    }

    private fun setHeaders(connection: HttpsURLConnection, headers: Map<String, String>) {
        for ((key, value) in headers) {
            connection.setRequestProperty(key, value)
        }
    }

    private fun sendBody(connection: HttpsURLConnection, model: RequestModel) {
        if (model.payload != null) {
            val payload = fromMap(model.payload!!.filterNotNull()).toString()
                .toByteArray(StandardCharsets.UTF_8)
            val writer = BufferedOutputStream(connection.outputStream)
            writer.write(payload)
            writer.close()
        }
    }

    private fun readResponse(connection: HttpsURLConnection?): ResponseModel {
        val statusCode = connection!!.responseCode
        val message = connection.responseMessage
        val headers = connection.headerFields
        val body = readBody(connection)
        return ResponseModel.Builder(timestampProvider)
            .statusCode(statusCode)
            .message(message)
            .headers(headers)
            .body(body)
            .requestModel(requestModel)
            .build()
    }

    private fun readBody(connection: HttpsURLConnection?): String {
        val responseCode = connection!!.responseCode
        val inputStream: InputStream = if (isStatusCodeOK(responseCode)) {
            connection.inputStream
        } else {
            connection.errorStream
        }
        val reader = BufferedReader(InputStreamReader(inputStream))
        var inputLine: String?
        val sb = StringBuilder()
        while (reader.readLine().also { inputLine = it } != null) {
            sb.append(inputLine)
        }
        reader.close()
        return sb.toString()
    }

    private fun isStatusCodeOK(responseCode: Int): Boolean {
        return responseCode in 200..299
    }
}